### ISSUE TITLE
Replace Highcharts with fully open-source charting library

### ISSUE THREAD

**Issue by brylie** [biolab/orange3#2044]
# Context
## Orange version
Orange 2.x and 3.x

## Expected behavior
Orange would comply with license terms of all internal components, and would **integrate only fully open-source components**.

## Actual behavior
The Highcharts library is **not fully open-source**.

## Additional info
Orange is *likely* in violation of the Highcharts license. Full details supporting this conclusion are available in a [discussion in the orange repository](https://github.com/biolab/orange/issues/45).

# Proposal
Replace Highcharts with a **fully open-source charting library**. Specifically, the following libraries are strong candidates:

- [Altair](https://altair-viz.github.io/) - Altair is a declarative statistical visualization library for Python, based on the powerful [Vega-Lite](http://vega.github.io/vega-lite) visualization grammar.
- [Bokeh](http://bokeh.pydata.org/en/latest/) - Bokeh is a Python interactive visualization library that targets modern web browsers for presentation.
- [Matplotlib D3 (mpld3)](http://mpld3.github.io/) - The mpld3 project brings together Matplotlib, the popular Python-based graphing library, and D3js, the popular JavaScript library for creating interactive data visualizations for the web.

**Comment by kernc** [comment]
Replacing Highcharts with Vega/Plotly. The add-ons that still use Highcharts have been relicensed to CC-BY-NC interim eventual porting. Making Vega nice and interactive looks _something_ to swallow.

A [Plotly](https://plot.ly/python/)-based implementation of parallel coordinates in https://github.com/biolab/orange3/pull/2327.

**Comment by brylie** [comment]
Does vega/vega-lite use CSS for styling? Vega seems to have a well concieved, mostly declarative, API.

Plot.ly does look nice. Are there any pros/cons to its Python/JS APIs ?

Also, which specific addons use Highcharts?

**Comment by brylie** [comment]
[Jake VanderPlas](http://www.vanderplas.com) (@jakevdp) gave a talk at PyCon 2017, entitled [The Python Visualization Landscape](https://www.youtube.com/watch?v=FytuB8nFHPQ). The presentation compares several major projects within the Python Visualization ecocystem, and how they relate. Projects directly relevant to this discussion include
- [Altair](https://altair-viz.github.io/) (part of [University of Washington eScience Institute](http://escience.washington.edu/) - @uwescience), 
- [Bokeh](http://bokeh.pydata.org/en/latest/) (from [Continuum Analytics](https://www.continuum.io/) - @ContinuumIO), and 
- [plot.ly](https://plot.ly/) (@plotly). 

[![pythonvisualizationlandscape-videothumb](https://cloud.githubusercontent.com/assets/17307/26307931/b68207ca-3eff-11e7-84c6-7f425adbacd8.jpg)](https://www.youtube.com/watch?v=FytuB8nFHPQ)

**Comment by kernc** [comment]
Vega / Altair lacks sufficient interactive examples, and interactivity is key in Orange. Besides, I've followed [the tutorial](https://vega.github.io/vega/tutorials/airports/) and was at the end trying to adapt their [Airports example](https://vega.github.io/vega-editor/?mode=vega&spec=airports) to weigh lines bolder by number of flights; took me a week to figure it out. Vega versions used in the tutorial and in the online editor aren't the same (3 vs 2) so copying code, as is implied in text, doesn't even work. Vega in whole felt like the visualization code will become, albeit safely packed in JSON, impossible to read afterwards and maintain in the long run. Instead of simple imperative JS event handlers anyone can reason about, you get safe-subset-of-JavaScript string expressions scattered all over the place.  The community is virtually non-existent ([_very_](https://stackoverflow.com/questions/tagged/vega) [few](https://stackoverflow.com/questions/tagged/bokeh) [questions](https://stackoverflow.com/questions/tagged/plotly) [on StackOverflow](https://stackoverflow.com/questions/tagged/highcharts) compared to competing libs, even though the project is nearly four years old); regardless of its purported eventual heavy use on Wikipedia, and it's "beautiful" declarative nature, I don't think this is something that will stick, or that complex interactive visualizations will be routinely built with it.

That's my current view of Vega. I have it wrapped, though, so it's in general trivial to include, should the need arise.

Bokes was a top contender, but Plotly seemed like a lighter weight solution and more like what we currently rely on. OTOH, for a completely new, browser-based solutions, I'd have no reservations about using [Bokeh controls for the UI](https://demo.bokehplots.com/apps/movies) as well.

> Does vega/vega-lite use CSS for styling?

These solutions mostly output SVG, so elements can be styled with CSS to a degree. Browsers' support is currently still limited to a subset of CSS directives (e.g. `text-overflow: ellipsis` for sure doesn't work).

> Plot.ly does look nice. Are there any pros/cons to its Python/JS APIs ?

JS API is JSON, Python API uses thin, arguments-checking dicts. It is declarative in the way of data and options passed. Easy to add JS event handlers. Probably harder to make completely custom visualizations, unless one is proficient in D3 anyway.

> Also, which specific addons use Highcharts?

Of public add-ons I know, [Timeseries](/biolab/orange3-timeseries) and [Educational](/biolab/orange3-educational).

**Comment by kernc** [comment]
> Vega seems to have a well concieved, mostly declarative, API.

Well-conceived? It's a flood of nested directions, few of which intuitively fit together and do exactly what one expects. Much like CSS, except that instead of 25M people knowing (about) it, 25 people do.

**Comment by brylie** [comment]
I just meant that the Vega grammar describes common characteristics of data visualization such as:

- [data](https://vega.github.io/vega/docs/data/)
- [transformations](https://vega.github.io/vega/docs/transforms/)
- [scales](https://vega.github.io/vega/docs/scales/)
- [axes](https://vega.github.io/vega/docs/axes/)
- [legends](https://vega.github.io/vega/docs/legends/)
- [markings](https://vega.github.io/vega/docs/marks)

The goal of the Vega specification is to help people think (declaratively) in terms of the visualizations they want to produce, rather than the (imperative) machine steps necessary to produce them.

@jakevdp, are there any examples of interactive visualizations made with Altair/Vega that we can use as reference?

**Comment by jakevdp** [comment]
FYI: vega/vega-lite just released a backward-incompatible update, and because I've been on parental leave I've not had time to update Altair, or update the JS library links that have changed. For the same reason, Altair doesn't currently support the interaction that was introduced a few weeks ago in the Vega-Lite release. Many of the difficulties you encountered are due to that, and I'll work on fixing them when I'm back to work in June.

I guess my request is: please don't judge the libraries on difficulties they present right now – it's quite temporary.

**Comment by jakevdp** [comment]
Same goes for documentation. I put some simple Altair docs out last fall, and decided to wait on putting concerted effort into them until the impeding vega-lite 2.0 release. It turns out that release, for various reasons, was gradually pushed back for about six months, so in retrospect I probably should have been working on the documentation in the meantime. Hindsight, 20/20, and all that.

**Comment by brylie** [comment]
Thank you so much for your response @jakevdp, and for your continued work on Altair. Regards to you and your family :smiley:

**Comment by domoritz** [comment]
@kernc Thank you for the interesting discussion. We are currently preparing the new releases and so some of the documentation is out of sync (we fixed the issue you pointed out in the docs at https://github.com/vega/vega/pull/821). You can try Vega 3 and Vega-Lite 2 in the online editor with those libraries installed. For example, at https://vega.github.io/new-editor/?mode=vega-lite&spec=brush&showEditor=1 you brush and select points (this example is Vega-Lite, the high-level language that compiles to Vega).

![brush](https://cloud.githubusercontent.com/assets/589034/26412938/391993dc-405f-11e7-86d8-2c40e002a965.gif)

Vega-Lite is more concise and so you wouldn't have to write (and maintain) large Vega specs. However, if you need to customize a visualization, you can always post-process the generated Vega spec. This is, in my option, really nice as it allows you to switch between abstraction levels as needed. 

We agree that the community is not that large but we have an active community at https://groups.google.com/forum/#!forum/vega-js. Our lab has built tools around authoring Vega and Vega-Lite and we will continue to work on these libraries. 

> Does vega/vega-lite use CSS for styling?

Vega-Lite compiles to Vega and Vega has different renders (SVG and Canvas). Because of this, you cannot reliably style with CSS. However, Vega supports all common styling options directly. For example, you can truncate text with `limit` (see example at https://vega.github.io/vega/docs/marks/text/).

**Comment by fab6** [comment]
I am far from being a expert, but bokeh seems to me quite powerful also in the long-run due to the continuum support.

**Comment by astaric** [comment]
Orange visualizations do not rely on highcharts (and never did). The issues was closed since highcharts libraries are no longer part of orange distribution so the licence compliance should be ok now.

There are two add-ons I am aware of that still rely on highcharts: [Timeseries](/biolab/orange3-timeseries) and [Educational](/biolab/orange3-educational). If anyone wants to start working on this, that would be IMHO the best place to start. Isuees can be opened there, where alternative libraries can be evaluated based on how well they are suited to reimplement existing visualizations.

If you would rather reuse this issue, feel free to reopen it.

**Comment by brylie** [comment]
OK, cool. I would just like to discuss more about using a Bokeh or Altair to build visualization widgets for Orange. So,it could be a separate issue, I reckon. Thanks for your reply.
